{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7580167d",
   "metadata": {},
   "source": [
    "# Compare total number of reads quantified/counted for different methods\n",
    "\n",
    "* htseq: we mapped reads against the genome and then counted genes using htseq. htseq counts genes that overlap regions in the genome defined by a GTF file. As will be shown later in this notebook, while the genomes had high mapping rates (not shown), the number of reads that counted toward a feature annotated in the GTF file was pretty low. This led us to explore other options, namely using the transcriptome to count reads.\n",
    "* salmon & a tx2gene file:  We observed higher mapping rates against the transcriptome, so we worked to create a `tx2gene` file, as differential expression analysis is generally more accurate at the gene level than the isoform level. We created this by using a splice-aware long read aligner (ultra) to map the assembled transcripts to the genome guided by the GTF file (uses minimap in the absence of GTF coordinates to splice on). We then assigned transcripts gene annotations by determining which genes they overlapped with using the mapping & GTF coordinates. While 91% of transcripts mapped, about half (approx 500k) transcripts were annotated as genes in this way. \n",
    "* salmon & a tx2gene file + transcripts with no gene annotations: We also augmented this tx2gene file with transcripts alone that didn't receive gene annotations. Presumably some fraction of reads map to these transcripts, so we back-added them to the tx2gene file to count these transcripts too.\n",
    "* salmon with transcripts only: This approach to counting should be redundant with the one above, but we included it just to check (spoiler, it's redundant)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fc5306d",
   "metadata": {},
   "source": [
    "## notbook setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c8fd170",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.2     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.2     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.1     \n",
      "── \u001b[1mConflicts\u001b[22m ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\n",
      "Attaching package: ‘jsonlite’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    flatten\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(tximport)\n",
    "library(jsonlite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "560dc439",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"..\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5235daf9",
   "metadata": {},
   "source": [
    "## functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "960a5bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_fastp_json <- function(file_path) {\n",
    "  # read the json file\n",
    "  json_data <- fromJSON(file_path)\n",
    "  \n",
    "  # extract num reads after filtering\n",
    "  total_reads_after_filtering <- json_data$summary$after_filtering$total_reads\n",
    "  \n",
    "  # return a data frame\n",
    "  return(data.frame(sample_name = gsub(\".json\", \"\", basename(file_path)),\n",
    "                    total_reads = total_reads_after_filtering))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7d54432",
   "metadata": {},
   "source": [
    "## read in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4781ef4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse fastp log files to get accurate total read counts after qc \n",
    "fastp <- Sys.glob(\"outputs/read_qc/fastp/*json\") %>%\n",
    "  map_dfr(read_fastp_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8341a1c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in htseq data frame and remove summary rows\n",
    "htseq <- read_tsv(\"outputs/counts/raw_counts.tsv\", show_col_types = F) %>%\n",
    "  filter(!grepl(\"^__\", gene)) # filter out summary rows produced by htseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "076833e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a vector of column names to use for all salmon counts\n",
    "salmon_colnames <- gsub(\"_quant\", \"\", basename(gsub(\"/quant.sf\", \"\", Sys.glob(\"outputs/quantification/*/quant.sf\"))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b3d2f050",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "10 \n",
      "11 \n",
      "12 \n",
      "13 \n",
      "14 \n",
      "15 \n",
      "16 \n",
      "17 \n",
      "18 \n",
      "19 \n",
      "20 \n",
      "21 \n",
      "22 \n",
      "23 \n",
      "24 \n",
      "25 \n",
      "26 \n",
      "27 \n",
      "28 \n",
      "29 \n",
      "30 \n",
      "31 \n",
      "32 \n",
      "33 \n",
      "34 \n",
      "35 \n",
      "36 \n",
      "37 \n",
      "38 \n",
      "39 \n",
      "40 \n",
      "41 \n",
      "42 \n",
      "43 \n",
      "44 \n",
      "45 \n",
      "46 \n",
      "47 \n",
      "48 \n",
      "49 \n",
      "50 \n",
      "51 \n",
      "52 \n",
      "53 \n",
      "54 \n",
      "55 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# read in salmon counts, output just the transcript counts without gene-level summarization\n",
    "salmon_tx <- tximport(files = Sys.glob(\"outputs/quantification/*/quant.sf\"),\n",
    "                      type = \"salmon\", txOut = TRUE)\n",
    "salmon_tx <- salmon_tx$counts\n",
    "colnames(salmon_tx) <- salmon_colnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2946be0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in the tx2gene file created by assigning transcripts gene names they overlapped with when mapped to to the reference genome\n",
    "# restore missing transcripts that don't have gene assignments and that didn't map to the genome\n",
    "tx2gene <- read_tsv(\"~/github/2023-amblyomma-americanum-txome-assembly/sandbox/try_ultra/Amblyomma_americanum_filtered_assembly.tx2gene_by_gtf.tsv\",\n",
    "                    skip = 1, col_names = c(\"tx\", \"gene\"), show_col_types = FALSE)\n",
    "missing_tx <- rownames(salmon_tx)[!rownames(salmon_tx) %in% tx2gene$tx]\n",
    "missing_tx <- data.frame(tx = missing_tx, gene = missing_tx)\n",
    "tx2gene <- bind_rows(tx2gene, missing_tx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bbebc2e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "10 \n",
      "11 \n",
      "12 \n",
      "13 \n",
      "14 \n",
      "15 \n",
      "16 \n",
      "17 \n",
      "18 \n",
      "19 \n",
      "20 \n",
      "21 \n",
      "22 \n",
      "23 \n",
      "24 \n",
      "25 \n",
      "26 \n",
      "27 \n",
      "28 \n",
      "29 \n",
      "30 \n",
      "31 \n",
      "32 \n",
      "33 \n",
      "34 \n",
      "35 \n",
      "36 \n",
      "37 \n",
      "38 \n",
      "39 \n",
      "40 \n",
      "41 \n",
      "42 \n",
      "43 \n",
      "44 \n",
      "45 \n",
      "46 \n",
      "47 \n",
      "48 \n",
      "49 \n",
      "50 \n",
      "51 \n",
      "52 \n",
      "53 \n",
      "54 \n",
      "55 \n",
      "\n",
      "\n",
      "summarizing abundance\n",
      "\n",
      "summarizing counts\n",
      "\n",
      "summarizing length\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# read in the salmon counts using the full tx2gene file with both genes and transcripts represented\n",
    "salmon_gx_tx <- tximport(files = Sys.glob(\"outputs/quantification/*/quant.sf\"),\n",
    "                         type = \"salmon\", txOut = FALSE, tx2gene = tx2gene)\n",
    "salmon_gx_tx <- salmon_gx_tx$counts\n",
    "colnames(salmon_gx_tx) <- salmon_colnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "904d5a6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter the tx2gene data.frame to only keep transcripts assigned to genes\n",
    "tx2gene_gx_only <- tx2gene %>%\n",
    "  filter(grepl(pattern = \"evm\", x = gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4f5c197b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "10 \n",
      "11 \n",
      "12 \n",
      "13 \n",
      "14 \n",
      "15 \n",
      "16 \n",
      "17 \n",
      "18 \n",
      "19 \n",
      "20 \n",
      "21 \n",
      "22 \n",
      "23 \n",
      "24 \n",
      "25 \n",
      "26 \n",
      "27 \n",
      "28 \n",
      "29 \n",
      "30 \n",
      "31 \n",
      "32 \n",
      "33 \n",
      "34 \n",
      "35 \n",
      "36 \n",
      "37 \n",
      "38 \n",
      "39 \n",
      "40 \n",
      "41 \n",
      "42 \n",
      "43 \n",
      "44 \n",
      "45 \n",
      "46 \n",
      "47 \n",
      "48 \n",
      "49 \n",
      "50 \n",
      "51 \n",
      "52 \n",
      "53 \n",
      "54 \n",
      "55 \n",
      "\n",
      "\n",
      "transcripts missing from tx2gene: 617596\n",
      "\n",
      "summarizing abundance\n",
      "\n",
      "summarizing counts\n",
      "\n",
      "summarizing length\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# read in salmon results and only keep counts that mapped to transcripts with gene assignments\n",
    "salmon_gx_only <- tximport(files = Sys.glob(\"outputs/quantification/*/quant.sf\"),\n",
    "                           type = \"salmon\", txOut = FALSE, \n",
    "                           tx2gene = tx2gene_gx_only)\n",
    "salmon_gx_only <- salmon_gx_only$counts\n",
    "colnames(salmon_gx_only) <- salmon_colnames"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9708d141",
   "metadata": {},
   "source": [
    "## calculate the total number of mapped reads for each method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "47087d24",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum_salmon_gx_only <- colSums(salmon_gx_only)\n",
    "sum_salmon_tx <- colSums(salmon_tx)\n",
    "sum_salmon_gx_tx <- colSums(salmon_gx_tx)\n",
    "sum_htseq <- colSums(htseq[ , -1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0dfead37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert results to data.frames\n",
    "mapping_sums_salmon <- data.frame(sample_name = names(sum_salmon_gx_only),\n",
    "                                  salmon_gx_only = sum_salmon_gx_only, \n",
    "                                  salmon_tx = sum_salmon_tx, \n",
    "                                  salmon_gx_tx = sum_salmon_gx_tx)\n",
    "mapping_sums_htseq <- data.frame(sample_name = names(sum_htseq),\n",
    "                                 htseq = sum_htseq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0072705d",
   "metadata": {},
   "source": [
    "## convert raw counts to percentages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "294e162d",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined <- left_join(fastp, mapping_sums_salmon, by = \"sample_name\") %>%\n",
    "  left_join(mapping_sums_htseq, by = \"sample_name\") %>%\n",
    "  mutate(p_salmon_gx_only = salmon_gx_only / total_reads,\n",
    "         p_salmon_gx_tx = salmon_gx_tx / total_reads,\n",
    "         p_salmon_tx = salmon_tx /total_reads,\n",
    "         p_htseq = htseq / total_reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "be1b6c23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 56 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>sample_name</th><th scope=col>total_reads</th><th scope=col>salmon_gx_only</th><th scope=col>salmon_tx</th><th scope=col>salmon_gx_tx</th><th scope=col>htseq</th><th scope=col>p_salmon_gx_only</th><th scope=col>p_salmon_gx_tx</th><th scope=col>p_salmon_tx</th><th scope=col>p_htseq</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AAFF          </td><td>   191306</td><td>   22065.41</td><td>   35707.01</td><td>   35707.01</td><td>   8101</td><td>0.11534090</td><td>0.18664868</td><td>0.18664868</td><td>0.04234577</td></tr>\n",
       "\t<tr><td>AAFM          </td><td>   101672</td><td>   16023.11</td><td>   20560.01</td><td>   20560.01</td><td>   6400</td><td>0.15759605</td><td>0.20221900</td><td>0.20221900</td><td>0.06294752</td></tr>\n",
       "\t<tr><td>AAUF          </td><td>   704018</td><td>   23508.16</td><td>   45493.01</td><td>   45493.01</td><td>   7301</td><td>0.03339142</td><td>0.06461909</td><td>0.06461909</td><td>0.01037047</td></tr>\n",
       "\t<tr><td>AAUM          </td><td>   156290</td><td>   38687.30</td><td>   49328.97</td><td>   49328.97</td><td>  11161</td><td>0.24753534</td><td>0.31562460</td><td>0.31562460</td><td>0.07141212</td></tr>\n",
       "\t<tr><td>AmbamSG7-11d  </td><td> 88372472</td><td>24349215.75</td><td>40773097.91</td><td>40773097.91</td><td>7131686</td><td>0.27552942</td><td>0.46137781</td><td>0.46137781</td><td>0.08070031</td></tr>\n",
       "\t<tr><td>AmbamSG72-144h</td><td> 99163790</td><td>26055582.29</td><td>45126079.96</td><td>45126079.96</td><td>7914394</td><td>0.26275299</td><td>0.45506611</td><td>0.45506611</td><td>0.07981133</td></tr>\n",
       "\t<tr><td>AmbamSGunfed  </td><td>104891872</td><td>22282481.44</td><td>45681707.96</td><td>45681707.96</td><td>6030717</td><td>0.21243287</td><td>0.43551237</td><td>0.43551237</td><td>0.05749461</td></tr>\n",
       "\t<tr><td>AmbameSG12-18h</td><td> 50175186</td><td>14832841.80</td><td>22922332.14</td><td>22922332.14</td><td>5236959</td><td>0.29562106</td><td>0.45684598</td><td>0.45684598</td><td>0.10437348</td></tr>\n",
       "\t<tr><td>Ecoli_12h_1   </td><td> 43216570</td><td>16769004.38</td><td>20688579.83</td><td>20688579.83</td><td>4725147</td><td>0.38802257</td><td>0.47871869</td><td>0.47871869</td><td>0.10933647</td></tr>\n",
       "\t<tr><td>Ecoli_12h_2   </td><td> 42471332</td><td>16523358.49</td><td>20353284.82</td><td>20353284.82</td><td>4622592</td><td>0.38904733</td><td>0.47922408</td><td>0.47922408</td><td>0.10884029</td></tr>\n",
       "\t<tr><td>Ecoli_12h_3   </td><td> 43699578</td><td>17414950.04</td><td>20938572.99</td><td>20938572.99</td><td>4922698</td><td>0.39851529</td><td>0.47914817</td><td>0.47914817</td><td>0.11264864</td></tr>\n",
       "\t<tr><td>Ecoli_24h_1   </td><td> 45076810</td><td>17952832.28</td><td>21497428.07</td><td>21497428.07</td><td>5119082</td><td>0.39827202</td><td>0.47690660</td><td>0.47690660</td><td>0.11356354</td></tr>\n",
       "\t<tr><td>Ecoli_24h_2   </td><td> 48230700</td><td>19027591.57</td><td>23043450.00</td><td>23043450.00</td><td>5453411</td><td>0.39451203</td><td>0.47777557</td><td>0.47777557</td><td>0.11306929</td></tr>\n",
       "\t<tr><td>Ecoli_24h_3   </td><td> 45399746</td><td>18024873.56</td><td>21631771.89</td><td>21631771.89</td><td>5121587</td><td>0.39702587</td><td>0.47647341</td><td>0.47647341</td><td>0.11281092</td></tr>\n",
       "\t<tr><td>Ecoli_3h_1    </td><td> 40090358</td><td>15308862.09</td><td>19087661.11</td><td>19087661.11</td><td>4215060</td><td>0.38185895</td><td>0.47611601</td><td>0.47611601</td><td>0.10513900</td></tr>\n",
       "\t<tr><td>Ecoli_3h_2    </td><td> 43793020</td><td>17265849.80</td><td>20896152.00</td><td>20896152.00</td><td>4874923</td><td>0.39426031</td><td>0.47715714</td><td>0.47715714</td><td>0.11131735</td></tr>\n",
       "\t<tr><td>Ecoli_3h_3    </td><td> 42786960</td><td>16239088.11</td><td>20421359.89</td><td>20421359.89</td><td>4491560</td><td>0.37953358</td><td>0.47727999</td><td>0.47727999</td><td>0.10497497</td></tr>\n",
       "\t<tr><td>Ecoli_6h_1    </td><td> 46461704</td><td>18065350.29</td><td>22090266.82</td><td>22090266.82</td><td>5042802</td><td>0.38882238</td><td>0.47545107</td><td>0.47545107</td><td>0.10853674</td></tr>\n",
       "\t<tr><td>Ecoli_6h_2    </td><td> 45089834</td><td>17803979.63</td><td>21500256.98</td><td>21500256.98</td><td>5083903</td><td>0.39485574</td><td>0.47683158</td><td>0.47683158</td><td>0.11275054</td></tr>\n",
       "\t<tr><td>Ecoli_6h_3    </td><td> 44228756</td><td>17555216.07</td><td>21047765.77</td><td>21047765.77</td><td>4966079</td><td>0.39691860</td><td>0.47588419</td><td>0.47588419</td><td>0.11228168</td></tr>\n",
       "\t<tr><td>If            </td><td> 35850183</td><td>25157606.56</td><td>32777543.94</td><td>32777543.94</td><td>7699015</td><td>0.70174277</td><td>0.91429223</td><td>0.91429223</td><td>0.21475525</td></tr>\n",
       "\t<tr><td>Im            </td><td> 35128509</td><td>22348360.39</td><td>28426840.86</td><td>28426840.86</td><td>7920166</td><td>0.63618870</td><td>0.80922424</td><td>0.80922424</td><td>0.22546263</td></tr>\n",
       "\t<tr><td>MG120         </td><td>   163654</td><td>   26573.19</td><td>   38086.99</td><td>   38086.99</td><td>   8377</td><td>0.16237420</td><td>0.23272874</td><td>0.23272874</td><td>0.05118726</td></tr>\n",
       "\t<tr><td>MG48          </td><td>   121880</td><td>   15597.03</td><td>   24801.99</td><td>   24801.99</td><td>   5208</td><td>0.12797040</td><td>0.20349515</td><td>0.20349515</td><td>0.04273055</td></tr>\n",
       "\t<tr><td>MG96          </td><td>    81342</td><td>   12767.30</td><td>   21062.00</td><td>   21062.00</td><td>   4096</td><td>0.15695827</td><td>0.25893139</td><td>0.25893139</td><td>0.05035529</td></tr>\n",
       "\t<tr><td>NEf           </td><td> 28366593</td><td>19675239.61</td><td>25855232.86</td><td>25855232.86</td><td>6166845</td><td>0.69360602</td><td>0.91146769</td><td>0.91146769</td><td>0.21739816</td></tr>\n",
       "\t<tr><td>NEm           </td><td> 39444686</td><td>25390406.42</td><td>32449822.93</td><td>32449822.93</td><td>8784247</td><td>0.64369650</td><td>0.82266653</td><td>0.82266653</td><td>0.22269786</td></tr>\n",
       "\t<tr><td>PBS_12h_1     </td><td> 47959866</td><td>19414124.68</td><td>23008433.91</td><td>23008433.91</td><td>5632622</td><td>0.40479939</td><td>0.47974350</td><td>0.47974350</td><td>0.11744449</td></tr>\n",
       "\t<tr><td>PBS_12h_2     </td><td> 44298492</td><td>18112614.11</td><td>21267395.96</td><td>21267395.96</td><td>5341489</td><td>0.40887654</td><td>0.48009300</td><td>0.48009300</td><td>0.12057948</td></tr>\n",
       "\t<tr><td>PBS_12h_3     </td><td> 45822366</td><td>18133816.14</td><td>21920882.93</td><td>21920882.93</td><td>5173217</td><td>0.39574159</td><td>0.47838828</td><td>0.47838828</td><td>0.11289720</td></tr>\n",
       "\t<tr><td>PBS_24h_1     </td><td> 43146760</td><td>17078459.46</td><td>20572145.92</td><td>20572145.92</td><td>4793091</td><td>0.39582252</td><td>0.47679469</td><td>0.47679469</td><td>0.11108809</td></tr>\n",
       "\t<tr><td>PBS_24h_2     </td><td> 50175276</td><td>19630503.41</td><td>23954096.91</td><td>23954096.91</td><td>5619071</td><td>0.39123857</td><td>0.47740837</td><td>0.47740837</td><td>0.11198884</td></tr>\n",
       "\t<tr><td>PBS_24h_3     </td><td> 47799116</td><td>18757223.24</td><td>22839891.70</td><td>22839891.70</td><td>5335866</td><td>0.39241779</td><td>0.47783084</td><td>0.47783084</td><td>0.11163106</td></tr>\n",
       "\t<tr><td>PBS_3h_1      </td><td> 47684372</td><td>         NA</td><td>         NA</td><td>         NA</td><td>     NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>PBS_3h_2      </td><td> 49840160</td><td>19617959.65</td><td>23765020.92</td><td>23765020.92</td><td>5385382</td><td>0.39361751</td><td>0.47682473</td><td>0.47682473</td><td>0.10805306</td></tr>\n",
       "\t<tr><td>PBS_3h_3      </td><td> 48322618</td><td>18994778.28</td><td>23098269.83</td><td>23098269.83</td><td>5363862</td><td>0.39308256</td><td>0.47800121</td><td>0.47800121</td><td>0.11100106</td></tr>\n",
       "\t<tr><td>PBS_6h_1      </td><td> 46799204</td><td>18843848.00</td><td>22393260.87</td><td>22393260.87</td><td>5412106</td><td>0.40265317</td><td>0.47849662</td><td>0.47849662</td><td>0.11564526</td></tr>\n",
       "\t<tr><td>PBS_6h_2      </td><td> 43940436</td><td>17513878.16</td><td>21033067.78</td><td>21033067.78</td><td>5084154</td><td>0.39858226</td><td>0.47867226</td><td>0.47867226</td><td>0.11570559</td></tr>\n",
       "\t<tr><td>PBS_6h_3      </td><td> 47693190</td><td>18746198.30</td><td>22706373.02</td><td>22706373.02</td><td>5433344</td><td>0.39305818</td><td>0.47609256</td><td>0.47609256</td><td>0.11392285</td></tr>\n",
       "\t<tr><td>PL1           </td><td>  1480408</td><td>  331257.32</td><td>  471380.01</td><td>  471380.01</td><td> 115426</td><td>0.22376083</td><td>0.31841223</td><td>0.31841223</td><td>0.07796905</td></tr>\n",
       "\t<tr><td>PL17          </td><td>   826838</td><td>  107263.34</td><td>  304418.00</td><td>  304418.00</td><td>  13472</td><td>0.12972715</td><td>0.36817128</td><td>0.36817128</td><td>0.01629340</td></tr>\n",
       "\t<tr><td>PL2           </td><td>  2061216</td><td>  444803.96</td><td>  679442.00</td><td>  679442.00</td><td> 110800</td><td>0.21579687</td><td>0.32963163</td><td>0.32963163</td><td>0.05375468</td></tr>\n",
       "\t<tr><td>PL3           </td><td>  1698968</td><td>  385310.61</td><td>  578938.98</td><td>  578938.98</td><td>  96227</td><td>0.22679097</td><td>0.34075920</td><td>0.34075920</td><td>0.05663850</td></tr>\n",
       "\t<tr><td>SG120         </td><td>   134436</td><td>   20417.18</td><td>   37765.01</td><td>   37765.01</td><td>   7123</td><td>0.15187285</td><td>0.28091440</td><td>0.28091440</td><td>0.05298432</td></tr>\n",
       "\t<tr><td>SG48          </td><td>   358690</td><td>   32859.12</td><td>   53500.01</td><td>   53500.01</td><td>  11243</td><td>0.09160869</td><td>0.14915388</td><td>0.14915388</td><td>0.03134462</td></tr>\n",
       "\t<tr><td>SG96          </td><td>   173112</td><td>   24887.28</td><td>   40225.00</td><td>   40225.00</td><td>   9303</td><td>0.14376403</td><td>0.23236401</td><td>0.23236401</td><td>0.05373978</td></tr>\n",
       "\t<tr><td>Solexa-205930 </td><td> 23988396</td><td> 9161988.68</td><td>11461137.01</td><td>11461137.01</td><td>2276070</td><td>0.38193419</td><td>0.47777838</td><td>0.47777838</td><td>0.09488213</td></tr>\n",
       "\t<tr><td>Solexa-205931 </td><td> 27014112</td><td>10313655.85</td><td>12921082.79</td><td>12921082.79</td><td>2549457</td><td>0.38178771</td><td>0.47830863</td><td>0.47830863</td><td>0.09437501</td></tr>\n",
       "\t<tr><td>Solexa-205932 </td><td> 30330478</td><td>11580119.27</td><td>14510001.94</td><td>14510001.94</td><td>2865981</td><td>0.38179811</td><td>0.47839674</td><td>0.47839674</td><td>0.09449178</td></tr>\n",
       "\t<tr><td>Solexa-205939 </td><td> 25526478</td><td> 9775432.59</td><td>12113243.97</td><td>12113243.97</td><td>2405526</td><td>0.38295266</td><td>0.47453644</td><td>0.47453644</td><td>0.09423650</td></tr>\n",
       "\t<tr><td>Solexa-205940 </td><td> 20327318</td><td> 7836897.54</td><td> 9710959.90</td><td> 9710959.90</td><td>1930365</td><td>0.38553525</td><td>0.47772952</td><td>0.47772952</td><td>0.09496408</td></tr>\n",
       "\t<tr><td>Solexa-205941 </td><td> 30002722</td><td>11521904.16</td><td>14319609.84</td><td>14319609.84</td><td>2855734</td><td>0.38402863</td><td>0.47727702</td><td>0.47727702</td><td>0.09518250</td></tr>\n",
       "\t<tr><td>Solexa-205942 </td><td> 28964952</td><td>11181984.97</td><td>13810595.03</td><td>13810595.03</td><td>2827339</td><td>0.38605225</td><td>0.47680366</td><td>0.47680366</td><td>0.09761242</td></tr>\n",
       "\t<tr><td>Solexa-206001 </td><td> 35687346</td><td>12784455.91</td><td>16514455.00</td><td>16514455.00</td><td>2936202</td><td>0.35823499</td><td>0.46275380</td><td>0.46275380</td><td>0.08227572</td></tr>\n",
       "\t<tr><td>Uf            </td><td> 33106695</td><td>24334310.07</td><td>30432591.76</td><td>30432591.76</td><td>7589434</td><td>0.73502686</td><td>0.91922772</td><td>0.91922772</td><td>0.22924167</td></tr>\n",
       "\t<tr><td>Um            </td><td> 36943285</td><td>22949303.44</td><td>28164174.77</td><td>28164174.77</td><td>8684513</td><td>0.62120365</td><td>0.76236249</td><td>0.76236249</td><td>0.23507690</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 56 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " sample\\_name & total\\_reads & salmon\\_gx\\_only & salmon\\_tx & salmon\\_gx\\_tx & htseq & p\\_salmon\\_gx\\_only & p\\_salmon\\_gx\\_tx & p\\_salmon\\_tx & p\\_htseq\\\\\n",
       " <chr> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t AAFF           &    191306 &    22065.41 &    35707.01 &    35707.01 &    8101 & 0.11534090 & 0.18664868 & 0.18664868 & 0.04234577\\\\\n",
       "\t AAFM           &    101672 &    16023.11 &    20560.01 &    20560.01 &    6400 & 0.15759605 & 0.20221900 & 0.20221900 & 0.06294752\\\\\n",
       "\t AAUF           &    704018 &    23508.16 &    45493.01 &    45493.01 &    7301 & 0.03339142 & 0.06461909 & 0.06461909 & 0.01037047\\\\\n",
       "\t AAUM           &    156290 &    38687.30 &    49328.97 &    49328.97 &   11161 & 0.24753534 & 0.31562460 & 0.31562460 & 0.07141212\\\\\n",
       "\t AmbamSG7-11d   &  88372472 & 24349215.75 & 40773097.91 & 40773097.91 & 7131686 & 0.27552942 & 0.46137781 & 0.46137781 & 0.08070031\\\\\n",
       "\t AmbamSG72-144h &  99163790 & 26055582.29 & 45126079.96 & 45126079.96 & 7914394 & 0.26275299 & 0.45506611 & 0.45506611 & 0.07981133\\\\\n",
       "\t AmbamSGunfed   & 104891872 & 22282481.44 & 45681707.96 & 45681707.96 & 6030717 & 0.21243287 & 0.43551237 & 0.43551237 & 0.05749461\\\\\n",
       "\t AmbameSG12-18h &  50175186 & 14832841.80 & 22922332.14 & 22922332.14 & 5236959 & 0.29562106 & 0.45684598 & 0.45684598 & 0.10437348\\\\\n",
       "\t Ecoli\\_12h\\_1    &  43216570 & 16769004.38 & 20688579.83 & 20688579.83 & 4725147 & 0.38802257 & 0.47871869 & 0.47871869 & 0.10933647\\\\\n",
       "\t Ecoli\\_12h\\_2    &  42471332 & 16523358.49 & 20353284.82 & 20353284.82 & 4622592 & 0.38904733 & 0.47922408 & 0.47922408 & 0.10884029\\\\\n",
       "\t Ecoli\\_12h\\_3    &  43699578 & 17414950.04 & 20938572.99 & 20938572.99 & 4922698 & 0.39851529 & 0.47914817 & 0.47914817 & 0.11264864\\\\\n",
       "\t Ecoli\\_24h\\_1    &  45076810 & 17952832.28 & 21497428.07 & 21497428.07 & 5119082 & 0.39827202 & 0.47690660 & 0.47690660 & 0.11356354\\\\\n",
       "\t Ecoli\\_24h\\_2    &  48230700 & 19027591.57 & 23043450.00 & 23043450.00 & 5453411 & 0.39451203 & 0.47777557 & 0.47777557 & 0.11306929\\\\\n",
       "\t Ecoli\\_24h\\_3    &  45399746 & 18024873.56 & 21631771.89 & 21631771.89 & 5121587 & 0.39702587 & 0.47647341 & 0.47647341 & 0.11281092\\\\\n",
       "\t Ecoli\\_3h\\_1     &  40090358 & 15308862.09 & 19087661.11 & 19087661.11 & 4215060 & 0.38185895 & 0.47611601 & 0.47611601 & 0.10513900\\\\\n",
       "\t Ecoli\\_3h\\_2     &  43793020 & 17265849.80 & 20896152.00 & 20896152.00 & 4874923 & 0.39426031 & 0.47715714 & 0.47715714 & 0.11131735\\\\\n",
       "\t Ecoli\\_3h\\_3     &  42786960 & 16239088.11 & 20421359.89 & 20421359.89 & 4491560 & 0.37953358 & 0.47727999 & 0.47727999 & 0.10497497\\\\\n",
       "\t Ecoli\\_6h\\_1     &  46461704 & 18065350.29 & 22090266.82 & 22090266.82 & 5042802 & 0.38882238 & 0.47545107 & 0.47545107 & 0.10853674\\\\\n",
       "\t Ecoli\\_6h\\_2     &  45089834 & 17803979.63 & 21500256.98 & 21500256.98 & 5083903 & 0.39485574 & 0.47683158 & 0.47683158 & 0.11275054\\\\\n",
       "\t Ecoli\\_6h\\_3     &  44228756 & 17555216.07 & 21047765.77 & 21047765.77 & 4966079 & 0.39691860 & 0.47588419 & 0.47588419 & 0.11228168\\\\\n",
       "\t If             &  35850183 & 25157606.56 & 32777543.94 & 32777543.94 & 7699015 & 0.70174277 & 0.91429223 & 0.91429223 & 0.21475525\\\\\n",
       "\t Im             &  35128509 & 22348360.39 & 28426840.86 & 28426840.86 & 7920166 & 0.63618870 & 0.80922424 & 0.80922424 & 0.22546263\\\\\n",
       "\t MG120          &    163654 &    26573.19 &    38086.99 &    38086.99 &    8377 & 0.16237420 & 0.23272874 & 0.23272874 & 0.05118726\\\\\n",
       "\t MG48           &    121880 &    15597.03 &    24801.99 &    24801.99 &    5208 & 0.12797040 & 0.20349515 & 0.20349515 & 0.04273055\\\\\n",
       "\t MG96           &     81342 &    12767.30 &    21062.00 &    21062.00 &    4096 & 0.15695827 & 0.25893139 & 0.25893139 & 0.05035529\\\\\n",
       "\t NEf            &  28366593 & 19675239.61 & 25855232.86 & 25855232.86 & 6166845 & 0.69360602 & 0.91146769 & 0.91146769 & 0.21739816\\\\\n",
       "\t NEm            &  39444686 & 25390406.42 & 32449822.93 & 32449822.93 & 8784247 & 0.64369650 & 0.82266653 & 0.82266653 & 0.22269786\\\\\n",
       "\t PBS\\_12h\\_1      &  47959866 & 19414124.68 & 23008433.91 & 23008433.91 & 5632622 & 0.40479939 & 0.47974350 & 0.47974350 & 0.11744449\\\\\n",
       "\t PBS\\_12h\\_2      &  44298492 & 18112614.11 & 21267395.96 & 21267395.96 & 5341489 & 0.40887654 & 0.48009300 & 0.48009300 & 0.12057948\\\\\n",
       "\t PBS\\_12h\\_3      &  45822366 & 18133816.14 & 21920882.93 & 21920882.93 & 5173217 & 0.39574159 & 0.47838828 & 0.47838828 & 0.11289720\\\\\n",
       "\t PBS\\_24h\\_1      &  43146760 & 17078459.46 & 20572145.92 & 20572145.92 & 4793091 & 0.39582252 & 0.47679469 & 0.47679469 & 0.11108809\\\\\n",
       "\t PBS\\_24h\\_2      &  50175276 & 19630503.41 & 23954096.91 & 23954096.91 & 5619071 & 0.39123857 & 0.47740837 & 0.47740837 & 0.11198884\\\\\n",
       "\t PBS\\_24h\\_3      &  47799116 & 18757223.24 & 22839891.70 & 22839891.70 & 5335866 & 0.39241779 & 0.47783084 & 0.47783084 & 0.11163106\\\\\n",
       "\t PBS\\_3h\\_1       &  47684372 &          NA &          NA &          NA &      NA &         NA &         NA &         NA &         NA\\\\\n",
       "\t PBS\\_3h\\_2       &  49840160 & 19617959.65 & 23765020.92 & 23765020.92 & 5385382 & 0.39361751 & 0.47682473 & 0.47682473 & 0.10805306\\\\\n",
       "\t PBS\\_3h\\_3       &  48322618 & 18994778.28 & 23098269.83 & 23098269.83 & 5363862 & 0.39308256 & 0.47800121 & 0.47800121 & 0.11100106\\\\\n",
       "\t PBS\\_6h\\_1       &  46799204 & 18843848.00 & 22393260.87 & 22393260.87 & 5412106 & 0.40265317 & 0.47849662 & 0.47849662 & 0.11564526\\\\\n",
       "\t PBS\\_6h\\_2       &  43940436 & 17513878.16 & 21033067.78 & 21033067.78 & 5084154 & 0.39858226 & 0.47867226 & 0.47867226 & 0.11570559\\\\\n",
       "\t PBS\\_6h\\_3       &  47693190 & 18746198.30 & 22706373.02 & 22706373.02 & 5433344 & 0.39305818 & 0.47609256 & 0.47609256 & 0.11392285\\\\\n",
       "\t PL1            &   1480408 &   331257.32 &   471380.01 &   471380.01 &  115426 & 0.22376083 & 0.31841223 & 0.31841223 & 0.07796905\\\\\n",
       "\t PL17           &    826838 &   107263.34 &   304418.00 &   304418.00 &   13472 & 0.12972715 & 0.36817128 & 0.36817128 & 0.01629340\\\\\n",
       "\t PL2            &   2061216 &   444803.96 &   679442.00 &   679442.00 &  110800 & 0.21579687 & 0.32963163 & 0.32963163 & 0.05375468\\\\\n",
       "\t PL3            &   1698968 &   385310.61 &   578938.98 &   578938.98 &   96227 & 0.22679097 & 0.34075920 & 0.34075920 & 0.05663850\\\\\n",
       "\t SG120          &    134436 &    20417.18 &    37765.01 &    37765.01 &    7123 & 0.15187285 & 0.28091440 & 0.28091440 & 0.05298432\\\\\n",
       "\t SG48           &    358690 &    32859.12 &    53500.01 &    53500.01 &   11243 & 0.09160869 & 0.14915388 & 0.14915388 & 0.03134462\\\\\n",
       "\t SG96           &    173112 &    24887.28 &    40225.00 &    40225.00 &    9303 & 0.14376403 & 0.23236401 & 0.23236401 & 0.05373978\\\\\n",
       "\t Solexa-205930  &  23988396 &  9161988.68 & 11461137.01 & 11461137.01 & 2276070 & 0.38193419 & 0.47777838 & 0.47777838 & 0.09488213\\\\\n",
       "\t Solexa-205931  &  27014112 & 10313655.85 & 12921082.79 & 12921082.79 & 2549457 & 0.38178771 & 0.47830863 & 0.47830863 & 0.09437501\\\\\n",
       "\t Solexa-205932  &  30330478 & 11580119.27 & 14510001.94 & 14510001.94 & 2865981 & 0.38179811 & 0.47839674 & 0.47839674 & 0.09449178\\\\\n",
       "\t Solexa-205939  &  25526478 &  9775432.59 & 12113243.97 & 12113243.97 & 2405526 & 0.38295266 & 0.47453644 & 0.47453644 & 0.09423650\\\\\n",
       "\t Solexa-205940  &  20327318 &  7836897.54 &  9710959.90 &  9710959.90 & 1930365 & 0.38553525 & 0.47772952 & 0.47772952 & 0.09496408\\\\\n",
       "\t Solexa-205941  &  30002722 & 11521904.16 & 14319609.84 & 14319609.84 & 2855734 & 0.38402863 & 0.47727702 & 0.47727702 & 0.09518250\\\\\n",
       "\t Solexa-205942  &  28964952 & 11181984.97 & 13810595.03 & 13810595.03 & 2827339 & 0.38605225 & 0.47680366 & 0.47680366 & 0.09761242\\\\\n",
       "\t Solexa-206001  &  35687346 & 12784455.91 & 16514455.00 & 16514455.00 & 2936202 & 0.35823499 & 0.46275380 & 0.46275380 & 0.08227572\\\\\n",
       "\t Uf             &  33106695 & 24334310.07 & 30432591.76 & 30432591.76 & 7589434 & 0.73502686 & 0.91922772 & 0.91922772 & 0.22924167\\\\\n",
       "\t Um             &  36943285 & 22949303.44 & 28164174.77 & 28164174.77 & 8684513 & 0.62120365 & 0.76236249 & 0.76236249 & 0.23507690\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 56 × 10\n",
       "\n",
       "| sample_name &lt;chr&gt; | total_reads &lt;int&gt; | salmon_gx_only &lt;dbl&gt; | salmon_tx &lt;dbl&gt; | salmon_gx_tx &lt;dbl&gt; | htseq &lt;dbl&gt; | p_salmon_gx_only &lt;dbl&gt; | p_salmon_gx_tx &lt;dbl&gt; | p_salmon_tx &lt;dbl&gt; | p_htseq &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| AAFF           |    191306 |    22065.41 |    35707.01 |    35707.01 |    8101 | 0.11534090 | 0.18664868 | 0.18664868 | 0.04234577 |\n",
       "| AAFM           |    101672 |    16023.11 |    20560.01 |    20560.01 |    6400 | 0.15759605 | 0.20221900 | 0.20221900 | 0.06294752 |\n",
       "| AAUF           |    704018 |    23508.16 |    45493.01 |    45493.01 |    7301 | 0.03339142 | 0.06461909 | 0.06461909 | 0.01037047 |\n",
       "| AAUM           |    156290 |    38687.30 |    49328.97 |    49328.97 |   11161 | 0.24753534 | 0.31562460 | 0.31562460 | 0.07141212 |\n",
       "| AmbamSG7-11d   |  88372472 | 24349215.75 | 40773097.91 | 40773097.91 | 7131686 | 0.27552942 | 0.46137781 | 0.46137781 | 0.08070031 |\n",
       "| AmbamSG72-144h |  99163790 | 26055582.29 | 45126079.96 | 45126079.96 | 7914394 | 0.26275299 | 0.45506611 | 0.45506611 | 0.07981133 |\n",
       "| AmbamSGunfed   | 104891872 | 22282481.44 | 45681707.96 | 45681707.96 | 6030717 | 0.21243287 | 0.43551237 | 0.43551237 | 0.05749461 |\n",
       "| AmbameSG12-18h |  50175186 | 14832841.80 | 22922332.14 | 22922332.14 | 5236959 | 0.29562106 | 0.45684598 | 0.45684598 | 0.10437348 |\n",
       "| Ecoli_12h_1    |  43216570 | 16769004.38 | 20688579.83 | 20688579.83 | 4725147 | 0.38802257 | 0.47871869 | 0.47871869 | 0.10933647 |\n",
       "| Ecoli_12h_2    |  42471332 | 16523358.49 | 20353284.82 | 20353284.82 | 4622592 | 0.38904733 | 0.47922408 | 0.47922408 | 0.10884029 |\n",
       "| Ecoli_12h_3    |  43699578 | 17414950.04 | 20938572.99 | 20938572.99 | 4922698 | 0.39851529 | 0.47914817 | 0.47914817 | 0.11264864 |\n",
       "| Ecoli_24h_1    |  45076810 | 17952832.28 | 21497428.07 | 21497428.07 | 5119082 | 0.39827202 | 0.47690660 | 0.47690660 | 0.11356354 |\n",
       "| Ecoli_24h_2    |  48230700 | 19027591.57 | 23043450.00 | 23043450.00 | 5453411 | 0.39451203 | 0.47777557 | 0.47777557 | 0.11306929 |\n",
       "| Ecoli_24h_3    |  45399746 | 18024873.56 | 21631771.89 | 21631771.89 | 5121587 | 0.39702587 | 0.47647341 | 0.47647341 | 0.11281092 |\n",
       "| Ecoli_3h_1     |  40090358 | 15308862.09 | 19087661.11 | 19087661.11 | 4215060 | 0.38185895 | 0.47611601 | 0.47611601 | 0.10513900 |\n",
       "| Ecoli_3h_2     |  43793020 | 17265849.80 | 20896152.00 | 20896152.00 | 4874923 | 0.39426031 | 0.47715714 | 0.47715714 | 0.11131735 |\n",
       "| Ecoli_3h_3     |  42786960 | 16239088.11 | 20421359.89 | 20421359.89 | 4491560 | 0.37953358 | 0.47727999 | 0.47727999 | 0.10497497 |\n",
       "| Ecoli_6h_1     |  46461704 | 18065350.29 | 22090266.82 | 22090266.82 | 5042802 | 0.38882238 | 0.47545107 | 0.47545107 | 0.10853674 |\n",
       "| Ecoli_6h_2     |  45089834 | 17803979.63 | 21500256.98 | 21500256.98 | 5083903 | 0.39485574 | 0.47683158 | 0.47683158 | 0.11275054 |\n",
       "| Ecoli_6h_3     |  44228756 | 17555216.07 | 21047765.77 | 21047765.77 | 4966079 | 0.39691860 | 0.47588419 | 0.47588419 | 0.11228168 |\n",
       "| If             |  35850183 | 25157606.56 | 32777543.94 | 32777543.94 | 7699015 | 0.70174277 | 0.91429223 | 0.91429223 | 0.21475525 |\n",
       "| Im             |  35128509 | 22348360.39 | 28426840.86 | 28426840.86 | 7920166 | 0.63618870 | 0.80922424 | 0.80922424 | 0.22546263 |\n",
       "| MG120          |    163654 |    26573.19 |    38086.99 |    38086.99 |    8377 | 0.16237420 | 0.23272874 | 0.23272874 | 0.05118726 |\n",
       "| MG48           |    121880 |    15597.03 |    24801.99 |    24801.99 |    5208 | 0.12797040 | 0.20349515 | 0.20349515 | 0.04273055 |\n",
       "| MG96           |     81342 |    12767.30 |    21062.00 |    21062.00 |    4096 | 0.15695827 | 0.25893139 | 0.25893139 | 0.05035529 |\n",
       "| NEf            |  28366593 | 19675239.61 | 25855232.86 | 25855232.86 | 6166845 | 0.69360602 | 0.91146769 | 0.91146769 | 0.21739816 |\n",
       "| NEm            |  39444686 | 25390406.42 | 32449822.93 | 32449822.93 | 8784247 | 0.64369650 | 0.82266653 | 0.82266653 | 0.22269786 |\n",
       "| PBS_12h_1      |  47959866 | 19414124.68 | 23008433.91 | 23008433.91 | 5632622 | 0.40479939 | 0.47974350 | 0.47974350 | 0.11744449 |\n",
       "| PBS_12h_2      |  44298492 | 18112614.11 | 21267395.96 | 21267395.96 | 5341489 | 0.40887654 | 0.48009300 | 0.48009300 | 0.12057948 |\n",
       "| PBS_12h_3      |  45822366 | 18133816.14 | 21920882.93 | 21920882.93 | 5173217 | 0.39574159 | 0.47838828 | 0.47838828 | 0.11289720 |\n",
       "| PBS_24h_1      |  43146760 | 17078459.46 | 20572145.92 | 20572145.92 | 4793091 | 0.39582252 | 0.47679469 | 0.47679469 | 0.11108809 |\n",
       "| PBS_24h_2      |  50175276 | 19630503.41 | 23954096.91 | 23954096.91 | 5619071 | 0.39123857 | 0.47740837 | 0.47740837 | 0.11198884 |\n",
       "| PBS_24h_3      |  47799116 | 18757223.24 | 22839891.70 | 22839891.70 | 5335866 | 0.39241779 | 0.47783084 | 0.47783084 | 0.11163106 |\n",
       "| PBS_3h_1       |  47684372 |          NA |          NA |          NA |      NA |         NA |         NA |         NA |         NA |\n",
       "| PBS_3h_2       |  49840160 | 19617959.65 | 23765020.92 | 23765020.92 | 5385382 | 0.39361751 | 0.47682473 | 0.47682473 | 0.10805306 |\n",
       "| PBS_3h_3       |  48322618 | 18994778.28 | 23098269.83 | 23098269.83 | 5363862 | 0.39308256 | 0.47800121 | 0.47800121 | 0.11100106 |\n",
       "| PBS_6h_1       |  46799204 | 18843848.00 | 22393260.87 | 22393260.87 | 5412106 | 0.40265317 | 0.47849662 | 0.47849662 | 0.11564526 |\n",
       "| PBS_6h_2       |  43940436 | 17513878.16 | 21033067.78 | 21033067.78 | 5084154 | 0.39858226 | 0.47867226 | 0.47867226 | 0.11570559 |\n",
       "| PBS_6h_3       |  47693190 | 18746198.30 | 22706373.02 | 22706373.02 | 5433344 | 0.39305818 | 0.47609256 | 0.47609256 | 0.11392285 |\n",
       "| PL1            |   1480408 |   331257.32 |   471380.01 |   471380.01 |  115426 | 0.22376083 | 0.31841223 | 0.31841223 | 0.07796905 |\n",
       "| PL17           |    826838 |   107263.34 |   304418.00 |   304418.00 |   13472 | 0.12972715 | 0.36817128 | 0.36817128 | 0.01629340 |\n",
       "| PL2            |   2061216 |   444803.96 |   679442.00 |   679442.00 |  110800 | 0.21579687 | 0.32963163 | 0.32963163 | 0.05375468 |\n",
       "| PL3            |   1698968 |   385310.61 |   578938.98 |   578938.98 |   96227 | 0.22679097 | 0.34075920 | 0.34075920 | 0.05663850 |\n",
       "| SG120          |    134436 |    20417.18 |    37765.01 |    37765.01 |    7123 | 0.15187285 | 0.28091440 | 0.28091440 | 0.05298432 |\n",
       "| SG48           |    358690 |    32859.12 |    53500.01 |    53500.01 |   11243 | 0.09160869 | 0.14915388 | 0.14915388 | 0.03134462 |\n",
       "| SG96           |    173112 |    24887.28 |    40225.00 |    40225.00 |    9303 | 0.14376403 | 0.23236401 | 0.23236401 | 0.05373978 |\n",
       "| Solexa-205930  |  23988396 |  9161988.68 | 11461137.01 | 11461137.01 | 2276070 | 0.38193419 | 0.47777838 | 0.47777838 | 0.09488213 |\n",
       "| Solexa-205931  |  27014112 | 10313655.85 | 12921082.79 | 12921082.79 | 2549457 | 0.38178771 | 0.47830863 | 0.47830863 | 0.09437501 |\n",
       "| Solexa-205932  |  30330478 | 11580119.27 | 14510001.94 | 14510001.94 | 2865981 | 0.38179811 | 0.47839674 | 0.47839674 | 0.09449178 |\n",
       "| Solexa-205939  |  25526478 |  9775432.59 | 12113243.97 | 12113243.97 | 2405526 | 0.38295266 | 0.47453644 | 0.47453644 | 0.09423650 |\n",
       "| Solexa-205940  |  20327318 |  7836897.54 |  9710959.90 |  9710959.90 | 1930365 | 0.38553525 | 0.47772952 | 0.47772952 | 0.09496408 |\n",
       "| Solexa-205941  |  30002722 | 11521904.16 | 14319609.84 | 14319609.84 | 2855734 | 0.38402863 | 0.47727702 | 0.47727702 | 0.09518250 |\n",
       "| Solexa-205942  |  28964952 | 11181984.97 | 13810595.03 | 13810595.03 | 2827339 | 0.38605225 | 0.47680366 | 0.47680366 | 0.09761242 |\n",
       "| Solexa-206001  |  35687346 | 12784455.91 | 16514455.00 | 16514455.00 | 2936202 | 0.35823499 | 0.46275380 | 0.46275380 | 0.08227572 |\n",
       "| Uf             |  33106695 | 24334310.07 | 30432591.76 | 30432591.76 | 7589434 | 0.73502686 | 0.91922772 | 0.91922772 | 0.22924167 |\n",
       "| Um             |  36943285 | 22949303.44 | 28164174.77 | 28164174.77 | 8684513 | 0.62120365 | 0.76236249 | 0.76236249 | 0.23507690 |\n",
       "\n"
      ],
      "text/plain": [
       "   sample_name    total_reads salmon_gx_only salmon_tx   salmon_gx_tx htseq  \n",
       "1  AAFF              191306      22065.41       35707.01    35707.01     8101\n",
       "2  AAFM              101672      16023.11       20560.01    20560.01     6400\n",
       "3  AAUF              704018      23508.16       45493.01    45493.01     7301\n",
       "4  AAUM              156290      38687.30       49328.97    49328.97    11161\n",
       "5  AmbamSG7-11d    88372472   24349215.75    40773097.91 40773097.91  7131686\n",
       "6  AmbamSG72-144h  99163790   26055582.29    45126079.96 45126079.96  7914394\n",
       "7  AmbamSGunfed   104891872   22282481.44    45681707.96 45681707.96  6030717\n",
       "8  AmbameSG12-18h  50175186   14832841.80    22922332.14 22922332.14  5236959\n",
       "9  Ecoli_12h_1     43216570   16769004.38    20688579.83 20688579.83  4725147\n",
       "10 Ecoli_12h_2     42471332   16523358.49    20353284.82 20353284.82  4622592\n",
       "11 Ecoli_12h_3     43699578   17414950.04    20938572.99 20938572.99  4922698\n",
       "12 Ecoli_24h_1     45076810   17952832.28    21497428.07 21497428.07  5119082\n",
       "13 Ecoli_24h_2     48230700   19027591.57    23043450.00 23043450.00  5453411\n",
       "14 Ecoli_24h_3     45399746   18024873.56    21631771.89 21631771.89  5121587\n",
       "15 Ecoli_3h_1      40090358   15308862.09    19087661.11 19087661.11  4215060\n",
       "16 Ecoli_3h_2      43793020   17265849.80    20896152.00 20896152.00  4874923\n",
       "17 Ecoli_3h_3      42786960   16239088.11    20421359.89 20421359.89  4491560\n",
       "18 Ecoli_6h_1      46461704   18065350.29    22090266.82 22090266.82  5042802\n",
       "19 Ecoli_6h_2      45089834   17803979.63    21500256.98 21500256.98  5083903\n",
       "20 Ecoli_6h_3      44228756   17555216.07    21047765.77 21047765.77  4966079\n",
       "21 If              35850183   25157606.56    32777543.94 32777543.94  7699015\n",
       "22 Im              35128509   22348360.39    28426840.86 28426840.86  7920166\n",
       "23 MG120             163654      26573.19       38086.99    38086.99     8377\n",
       "24 MG48              121880      15597.03       24801.99    24801.99     5208\n",
       "25 MG96               81342      12767.30       21062.00    21062.00     4096\n",
       "26 NEf             28366593   19675239.61    25855232.86 25855232.86  6166845\n",
       "27 NEm             39444686   25390406.42    32449822.93 32449822.93  8784247\n",
       "28 PBS_12h_1       47959866   19414124.68    23008433.91 23008433.91  5632622\n",
       "29 PBS_12h_2       44298492   18112614.11    21267395.96 21267395.96  5341489\n",
       "30 PBS_12h_3       45822366   18133816.14    21920882.93 21920882.93  5173217\n",
       "31 PBS_24h_1       43146760   17078459.46    20572145.92 20572145.92  4793091\n",
       "32 PBS_24h_2       50175276   19630503.41    23954096.91 23954096.91  5619071\n",
       "33 PBS_24h_3       47799116   18757223.24    22839891.70 22839891.70  5335866\n",
       "34 PBS_3h_1        47684372            NA             NA          NA       NA\n",
       "35 PBS_3h_2        49840160   19617959.65    23765020.92 23765020.92  5385382\n",
       "36 PBS_3h_3        48322618   18994778.28    23098269.83 23098269.83  5363862\n",
       "37 PBS_6h_1        46799204   18843848.00    22393260.87 22393260.87  5412106\n",
       "38 PBS_6h_2        43940436   17513878.16    21033067.78 21033067.78  5084154\n",
       "39 PBS_6h_3        47693190   18746198.30    22706373.02 22706373.02  5433344\n",
       "40 PL1              1480408     331257.32      471380.01   471380.01   115426\n",
       "41 PL17              826838     107263.34      304418.00   304418.00    13472\n",
       "42 PL2              2061216     444803.96      679442.00   679442.00   110800\n",
       "43 PL3              1698968     385310.61      578938.98   578938.98    96227\n",
       "44 SG120             134436      20417.18       37765.01    37765.01     7123\n",
       "45 SG48              358690      32859.12       53500.01    53500.01    11243\n",
       "46 SG96              173112      24887.28       40225.00    40225.00     9303\n",
       "47 Solexa-205930   23988396    9161988.68    11461137.01 11461137.01  2276070\n",
       "48 Solexa-205931   27014112   10313655.85    12921082.79 12921082.79  2549457\n",
       "49 Solexa-205932   30330478   11580119.27    14510001.94 14510001.94  2865981\n",
       "50 Solexa-205939   25526478    9775432.59    12113243.97 12113243.97  2405526\n",
       "51 Solexa-205940   20327318    7836897.54     9710959.90  9710959.90  1930365\n",
       "52 Solexa-205941   30002722   11521904.16    14319609.84 14319609.84  2855734\n",
       "53 Solexa-205942   28964952   11181984.97    13810595.03 13810595.03  2827339\n",
       "54 Solexa-206001   35687346   12784455.91    16514455.00 16514455.00  2936202\n",
       "55 Uf              33106695   24334310.07    30432591.76 30432591.76  7589434\n",
       "56 Um              36943285   22949303.44    28164174.77 28164174.77  8684513\n",
       "   p_salmon_gx_only p_salmon_gx_tx p_salmon_tx p_htseq   \n",
       "1  0.11534090       0.18664868     0.18664868  0.04234577\n",
       "2  0.15759605       0.20221900     0.20221900  0.06294752\n",
       "3  0.03339142       0.06461909     0.06461909  0.01037047\n",
       "4  0.24753534       0.31562460     0.31562460  0.07141212\n",
       "5  0.27552942       0.46137781     0.46137781  0.08070031\n",
       "6  0.26275299       0.45506611     0.45506611  0.07981133\n",
       "7  0.21243287       0.43551237     0.43551237  0.05749461\n",
       "8  0.29562106       0.45684598     0.45684598  0.10437348\n",
       "9  0.38802257       0.47871869     0.47871869  0.10933647\n",
       "10 0.38904733       0.47922408     0.47922408  0.10884029\n",
       "11 0.39851529       0.47914817     0.47914817  0.11264864\n",
       "12 0.39827202       0.47690660     0.47690660  0.11356354\n",
       "13 0.39451203       0.47777557     0.47777557  0.11306929\n",
       "14 0.39702587       0.47647341     0.47647341  0.11281092\n",
       "15 0.38185895       0.47611601     0.47611601  0.10513900\n",
       "16 0.39426031       0.47715714     0.47715714  0.11131735\n",
       "17 0.37953358       0.47727999     0.47727999  0.10497497\n",
       "18 0.38882238       0.47545107     0.47545107  0.10853674\n",
       "19 0.39485574       0.47683158     0.47683158  0.11275054\n",
       "20 0.39691860       0.47588419     0.47588419  0.11228168\n",
       "21 0.70174277       0.91429223     0.91429223  0.21475525\n",
       "22 0.63618870       0.80922424     0.80922424  0.22546263\n",
       "23 0.16237420       0.23272874     0.23272874  0.05118726\n",
       "24 0.12797040       0.20349515     0.20349515  0.04273055\n",
       "25 0.15695827       0.25893139     0.25893139  0.05035529\n",
       "26 0.69360602       0.91146769     0.91146769  0.21739816\n",
       "27 0.64369650       0.82266653     0.82266653  0.22269786\n",
       "28 0.40479939       0.47974350     0.47974350  0.11744449\n",
       "29 0.40887654       0.48009300     0.48009300  0.12057948\n",
       "30 0.39574159       0.47838828     0.47838828  0.11289720\n",
       "31 0.39582252       0.47679469     0.47679469  0.11108809\n",
       "32 0.39123857       0.47740837     0.47740837  0.11198884\n",
       "33 0.39241779       0.47783084     0.47783084  0.11163106\n",
       "34         NA               NA             NA          NA\n",
       "35 0.39361751       0.47682473     0.47682473  0.10805306\n",
       "36 0.39308256       0.47800121     0.47800121  0.11100106\n",
       "37 0.40265317       0.47849662     0.47849662  0.11564526\n",
       "38 0.39858226       0.47867226     0.47867226  0.11570559\n",
       "39 0.39305818       0.47609256     0.47609256  0.11392285\n",
       "40 0.22376083       0.31841223     0.31841223  0.07796905\n",
       "41 0.12972715       0.36817128     0.36817128  0.01629340\n",
       "42 0.21579687       0.32963163     0.32963163  0.05375468\n",
       "43 0.22679097       0.34075920     0.34075920  0.05663850\n",
       "44 0.15187285       0.28091440     0.28091440  0.05298432\n",
       "45 0.09160869       0.14915388     0.14915388  0.03134462\n",
       "46 0.14376403       0.23236401     0.23236401  0.05373978\n",
       "47 0.38193419       0.47777838     0.47777838  0.09488213\n",
       "48 0.38178771       0.47830863     0.47830863  0.09437501\n",
       "49 0.38179811       0.47839674     0.47839674  0.09449178\n",
       "50 0.38295266       0.47453644     0.47453644  0.09423650\n",
       "51 0.38553525       0.47772952     0.47772952  0.09496408\n",
       "52 0.38402863       0.47727702     0.47727702  0.09518250\n",
       "53 0.38605225       0.47680366     0.47680366  0.09761242\n",
       "54 0.35823499       0.46275380     0.46275380  0.08227572\n",
       "55 0.73502686       0.91922772     0.91922772  0.22924167\n",
       "56 0.62120365       0.76236249     0.76236249  0.23507690"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "combined"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b25d234",
   "metadata": {},
   "source": [
    "## calculate mean mapping percentage for each method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4a2fe734",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>total_reads</dt><dd>33492252.27</dd><dt>salmon_gx_only</dt><dd>13196875.76</dd><dt>salmon_tx</dt><dd>17093211.08</dd><dt>salmon_gx_tx</dt><dd>17093211.08</dd><dt>htseq</dt><dd>3844217.02</dd><dt>p_salmon_gx_only</dt><dd>0.35</dd><dt>p_salmon_gx_tx</dt><dd>0.46</dd><dt>p_salmon_tx</dt><dd>0.46</dd><dt>p_htseq</dt><dd>0.1</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[total\\textbackslash{}\\_reads] 33492252.27\n",
       "\\item[salmon\\textbackslash{}\\_gx\\textbackslash{}\\_only] 13196875.76\n",
       "\\item[salmon\\textbackslash{}\\_tx] 17093211.08\n",
       "\\item[salmon\\textbackslash{}\\_gx\\textbackslash{}\\_tx] 17093211.08\n",
       "\\item[htseq] 3844217.02\n",
       "\\item[p\\textbackslash{}\\_salmon\\textbackslash{}\\_gx\\textbackslash{}\\_only] 0.35\n",
       "\\item[p\\textbackslash{}\\_salmon\\textbackslash{}\\_gx\\textbackslash{}\\_tx] 0.46\n",
       "\\item[p\\textbackslash{}\\_salmon\\textbackslash{}\\_tx] 0.46\n",
       "\\item[p\\textbackslash{}\\_htseq] 0.1\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "total_reads\n",
       ":   33492252.27salmon_gx_only\n",
       ":   13196875.76salmon_tx\n",
       ":   17093211.08salmon_gx_tx\n",
       ":   17093211.08htseq\n",
       ":   3844217.02p_salmon_gx_only\n",
       ":   0.35p_salmon_gx_tx\n",
       ":   0.46p_salmon_tx\n",
       ":   0.46p_htseq\n",
       ":   0.1\n",
       "\n"
      ],
      "text/plain": [
       "     total_reads   salmon_gx_only        salmon_tx     salmon_gx_tx \n",
       "     33492252.27      13196875.76      17093211.08      17093211.08 \n",
       "           htseq p_salmon_gx_only   p_salmon_gx_tx      p_salmon_tx \n",
       "      3844217.02             0.35             0.46             0.46 \n",
       "         p_htseq \n",
       "            0.10 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "round(colMeans(combined[ , -1], na.rm = T), digits = 2)\n",
    "\n",
    "# the columns that start with `p_*` are percents (actually fractions) for each counting strategy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "204e71ae",
   "metadata": {},
   "source": [
    "## summary\n",
    "\n",
    "As can be seen from the preceding cell, the percent of reads mapped by salmon, no matter which `tx2gene` file we use, is on average 3x higher than when we map to the genome and count reads that fall in feature regions.\n",
    "\n",
    "Surprisingly, there's only an average of 11% gain by mapping to all transcripts (500k _more_ transcripts than is in the gene-only tx2gene file) than to just the gene-only transcripts.\n",
    "\n",
    "Given these results, I'm leaning toward using the salmon counts with only the transcripts that mapped to genes represented. I don't think its worth the extra noise to include 500k more transcripts to gain 11% more read mapping. Note that if my de novo clustering (mmseqs2) had been more successful, I would have clustered these transcripts and then included those cluster representatives in my tx2gene file. However I got very little reduction with clustering (never dropped below 1M transcripts, even when dropping my percent identity to 70%), so I'm just going to ignore those transcripts for now."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96bbd3c7",
   "metadata": {},
   "source": [
    "## print session info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c077e700",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.0 (2023-04-21)\n",
       "Platform: x86_64-apple-darwin13.4.0 (64-bit)\n",
       "Running under: macOS Big Sur ... 10.16\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /Users/taylorreiter/miniconda3/envs/tidyjupyter/lib/libopenblasp-r0.3.23.dylib;  LAPACK version 3.11.0\n",
       "\n",
       "locale:\n",
       "[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8\n",
       "\n",
       "time zone: America/New_York\n",
       "tzcode source: system (macOS)\n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] jsonlite_1.8.7  tximport_1.28.0 lubridate_1.9.2 forcats_1.0.0  \n",
       " [5] stringr_1.5.0   dplyr_1.1.2     purrr_1.0.1     readr_2.1.4    \n",
       " [9] tidyr_1.3.0     tibble_3.2.1    ggplot2_3.4.2   tidyverse_2.0.0\n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] bit_4.0.5        gtable_0.3.3     compiler_4.3.0   crayon_1.5.2    \n",
       " [5] tidyselect_1.2.0 IRdisplay_1.1    parallel_4.3.0   scales_1.2.1    \n",
       " [9] uuid_1.1-0       fastmap_1.1.1    IRkernel_1.3.2   R6_2.5.1        \n",
       "[13] generics_0.1.3   munsell_0.5.0    pillar_1.9.0     tzdb_0.4.0      \n",
       "[17] rlang_1.1.1      utf8_1.2.3       stringi_1.7.12   repr_1.1.6      \n",
       "[21] bit64_4.0.5      timechange_0.2.0 cli_3.6.1        withr_2.5.0     \n",
       "[25] magrittr_2.0.3   digest_0.6.31    grid_4.3.0       vroom_1.6.3     \n",
       "[29] base64enc_0.1-3  hms_1.1.3        pbdZMQ_0.3-9     lifecycle_1.0.3 \n",
       "[33] vctrs_0.6.3      evaluate_0.21    glue_1.6.2       fansi_1.0.4     \n",
       "[37] colorspace_2.1-0 tools_4.3.0      pkgconfig_2.0.3  htmltools_0.5.5 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1065ed1b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
